BUG: .get_indexer_non_unique() must return an array of ints (#44084)

johannes-mueller opened the pull request:
GH#44084 boils down to the following.

According to the docs `.get_indexer_non_unique()` is supposed to return
"integers from 0 to n - 1 indicating that the index at these positions matches
the corresponding target values".  However, for an index that is non unique and
non monotonic it returns a boolean mask.  That is because it uses `.get_loc()`
which for non unique, non monotonic indexes returns a boolean mask.

This patch catches that case and converts the boolean mask from `.get_loc()`
into the corresponding array of integers if the index is not unique and not
monotonic.

- [x] closes #44084
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

johannes-mueller committed:
BUG: .get_indexer_non_unique() must return an array of ints (#44084)

GH#44084 boils down to the following.

According to the docs `.get_indexer_non_unique()` is supposed to return
"integers from 0 to n - 1 indicating that the index at these positions matches
the corresponding target values".  However, for an index that is non unique and
non monotonic it returns a boolean mask.  That is because it uses `.get_loc()`
which for non unique, non monotonic indexes returns a boolean mask.

This patch catches that case and converts the boolean mask from `.get_loc()`
into the corresponding array of integers if the index is not unique and not
monotonic.

phofl commented on pandas/tests/indexes/interval/test_indexing.py:
```diff
@@ -373,6 +373,16 @@ def test_get_indexer_with_nans(self):
 
+    def test_get_index_non_unique_non_monotonic(self):
+        # GH#44084
+        index = IntervalIndex.from_tuples(
```
Could you add the MultiIndex cases

phofl reviewed (changes requested):
Please add whatsnew

johannes-mueller committed:
Add whatsnew entry about fix for (#44084)

johannes-mueller committed:
Add MultiIndex test case, as shown in the bug report (#44084)

johannes-mueller committed:
Explicitly expect int64 arrays from .get_indexer_non_unique()

On Windows `np.array([1, 3])` is obviously int32 and thus the comparison to the
int64 array fails due to dtype mismatch.

johannes-mueller committed:
Correctly choose the int dtype to be expected from indexing methods

Sometimes the world out there is a bit more complicated than what you have on
your cozy desktop :)

jreback commented:
thanks @johannes-mueller 

do we have sufficient testing on this for *other* index types aside from those explicityly tested here? if not would take a PR for that
